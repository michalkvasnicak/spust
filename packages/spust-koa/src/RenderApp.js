// @flow

import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

import Middleware from './Middleware';

type Base = {
  href?: string,
  target?: string,
};
type Link = {
  crossOrigin?: string,
  href?: string,
  hrefLang?: string,
  integrity?: string,
  media?: string,
  preload?: boolean,
  prefetch?: boolean,
  rel?: string,
  sizes?: string,
  title?: string,
  type?: string,
};
type Meta = { content?: string, httpEquiv?: string, charSet?: string, name?: string };
type Script = {
  async?: boolean,
  crossOrigin?: string,
  defer?: boolean,
  integrity?: string,
  script?: string,
  src?: string,
  type?: string,
};
type Stylesheet = { href?: string, media?: string, rel?: string };

type Head = {
  base?: Base,
  meta?: Array<Meta>,
  links?: Array<Link>,
  scripts?: Array<Script>,
  stylesheets?: Array<Stylesheet>,
  title?: string,
};

type Result = {
  body?: string,
  containerId?: string,
  head?: Head,
  lang?: string,
  scripts?: Array<Script>,
  status?: number,
  stylesheets?: Array<Stylesheet>,
};

export type Props = {
  render: (context: *) => Promise<Result> | Result,
};

export function renderHead(head: Head, additionalStylesheets?: Array<Stylesheet> = []): string {
  const metaTags = head.meta || [];
  const links = head.links || [];
  const scripts = head.scripts || [];
  const stylesheets = [...(head.stylesheets || []), ...additionalStylesheets];

  return renderToStaticMarkup(
    <head>
      {head.base && <base {...head.base} />}
      {metaTags.map((tag, i) => <meta key={i} {...tag} />)}
      <title>{head.title}</title>
      {links.map((linkAttrs, i) => <link key={i} {...linkAttrs} />)}
      {stylesheets.map((stylesheetAttrs, i) =>
        <link key={`s-${i}`} {...stylesheetAttrs} rel={stylesheetAttrs.rel || 'stylesheet'} />,
      )}
      {scripts.map((scriptAttrs, i) =>
        <script
          key={`scr-${i}`}
          {...scriptAttrs}
          dangerouslySetInnerHTML={{ __html: scriptAttrs.script }}
        />,
      )}
    </head>,
  );
}

export function renderFooter(scripts?: Array<Script> = []): string {
  return renderToStaticMarkup(
    <footer>
      {scripts.map((scriptAttrs, i) =>
        <script
          key={`fscr-${i}`}
          {...scriptAttrs}
          dangerouslySetInnerHTML={{ __html: scriptAttrs.script }}
        />,
      )}
    </footer>,
  ).replace(/<(\/)?footer>/g, '');
}

export default function RenderApp({ render }: Props) {
  return (
    <Middleware
      use={async ctx => {
        const {
          body = '',
          containerId = 'app',
          lang = 'en',
          head = {},
          scripts = [],
          status,
          stylesheets = [],
        } = await render(ctx);

        ctx.status = status || 200;

        ctx.body = `
<html lang="${lang}">
  ${renderHead(head, stylesheets)}
  <body>
    <div id="${containerId}">${body}</div>
    ${renderFooter(scripts)}
  </body>
</html>
    `.trim();
      }}
    />
  );
}
