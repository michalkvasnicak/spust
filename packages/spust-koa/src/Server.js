// @flow

import Koa from 'koa';
import PropTypes from 'prop-types';
import React from 'react';
import { Server as HttpServer } from 'http';

// eslint-disable-next-line no-unused-vars
type ExtractMiddlewareType<MT: *, T: (mw: MT) => *> = MT;

export type MiddlewareFn = ExtractMiddlewareType<any, $PropertyType<Koa, 'use'>>;

export type ServerContext = {
  app: Koa,
  listen: (...args: any) => HttpServer,
};

export type Props = {
  context: Object,
  children?: any,
  port: number,
};

type DefaultProps = {
  context: Object,
};

export type Context = {
  server: Koa,
  use: $PropertyType<Koa, 'use'>,
};

export const serverContextType = {
  use: PropTypes.func.isRequired,
  server: PropTypes.instanceOf(Koa).isRequired,
};

export default class Server extends React.Component<DefaultProps, Props, void> {
  app: Koa;

  static defaultProps = {
    context: {},
  };

  static childContextTypes = serverContextType;

  constructor(props: Props, context: any) {
    super(props, context);

    this.app = new Koa();

    this.props.context.app = this.app;
    this.props.context.listen = () => this.app.listen(this.props.port);
  }

  getChildContext = () => ({
    use: (middleware: MiddlewareFn) => this.app.use(middleware),
    server: this.app,
  });

  render() {
    return (
      <div>
        {this.props.children}
      </div>
    );
  }
}
