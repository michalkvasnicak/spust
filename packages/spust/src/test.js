#!/usr/bin/env node
// @flow

process.env.NODE_ENV = 'test';
process.env.BABEL_ENV = 'test';

process.on('unhandledRejection', err => {
  throw err;
});

const jest = require('jest');
const fs = require('fs');
const path = require('path');
const argv = process.argv.slice(2);

if (!process.env.CI && argv.indexOf('--coverage') < 0) {
  argv.push('--watch');
} else {
  argv.push('--forceExit');
}

const setupTestsFile = fs.existsSync(path.resolve(process.cwd(), 'src/setupTests.js'))
  ? '<rootDir>/src/setupTests.js'
  : undefined;

argv.push(
  '--config',
  JSON.stringify({
    collectCoverageFrom: ['src/**/*.{js,jsx}'],
    setupFiles: [path.resolve(__dirname, './polyfills/server.js')],
    setupTestFrameworkScriptFile: setupTestsFile,
    testMatch: [
      '<rootDir>/src/**/__tests__/**/*.js?(x)',
      '<rootDir>/src/**/?(*.)(spec|test).js?(x)',
    ],
    testEnvironment: 'node',
    testURL: 'http://localhost',
    transform: {
      '^.+\\.(js|jsx)$': path.resolve(__dirname, 'jest/babelTransform.js'),
      '^.+\\.css$': path.resolve(__dirname, 'jest/cssTransform.js'),
      '^(?!.*\\.(js|jsx|css|json)$)': path.resolve(__dirname, 'jest/fileTransform.js'),
    },
    transformIgnorePatterns: ['[/\\\\]node_modules[/\\\\].+\\.(js|jsx)$'],
  }),
);

jest.run(argv);
