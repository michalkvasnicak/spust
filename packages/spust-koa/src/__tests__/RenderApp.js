// @flow

import React from 'react';
import test from 'supertest';

import RenderApp from '../RenderApp';
import Server from '../Server';
import serve from '../';

describe('RenderApp', () => {
  it('works correctly', async () => {
    const server = serve(
      <Server port={3000}>
        <RenderApp
          render={ctx => ({
            head: {
              links: [{ crossOrigin: 'same-origin', href: 'test', rel: 'stylesheet' }],
              meta: [{ content: 'test-value', name: 'test' }],
              stylesheets: [{ href: 'test2' }],
            },
          })}
        />
      </Server>,
      false,
    );

    await test(server.app.callback())
      .options('/')
      .expect(200)
      .expect(res => expect(res.text).toMatchSnapshot());
  });
});
