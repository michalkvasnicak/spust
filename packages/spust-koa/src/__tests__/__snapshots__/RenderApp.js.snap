// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderApp works correctly 1`] = `
"<html lang=\\"en\\">
  <head><meta content=\\"test-value\\" name=\\"test\\"/><title></title><link crossorigin=\\"same-origin\\" href=\\"test\\" rel=\\"stylesheet\\"/><link href=\\"test2\\" rel=\\"stylesheet\\"/></head>
  <body>
    <div id=\\"app\\"></div>
    
  </body>
</html>"
`;
